dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1002 @@
+          resource.zone = 'global'
+
+          # For creation of a snapshot, its a special case where the targetId is
+          # of the source disk and not the snapshot being created. And the
+          # targetLink points to the source disk. The id of the snapshot
+          # is not present in the log message.
+          # We need to unset the id which was set earlier to targetId
+          matched_resource = self._FindResource(resource.name, resource.type)
+          if matched_resource:
+            resource = matched_resource
+          else:
+            resource.id = source_disk_id + "-snapshot"
+
+        else:
+          resource.resource_name = resource.resource_name or proto_payload.get(
+              'resourceName', '')
+
+        # In case the message is an insert message
+        if log_message_type.endswith('insert') or log_message_type.endswith(
+            'createSnapshot'):
+          if request_metadata:
+            resource.creator_ip_address = request_metadata.get('callerIp')
+            resource.creator_useragent = request_metadata.get(
+                'callerSuppliedUserAgent')
+          self._ParseInsertLogMessage(resource, request, response)
+
+        elif log_message_type.endswith('delete'):
+          resource.deletion_timestamp = response.get('insertTime')
+          resource.deleted_by = response.get('user')
+          if request_metadata:
+            resource.deleter_ip_address = request_metadata.get('callerIp')
+            resource.deleter_useragent = request_metadata.get(
+                'callerSuppliedUserAgent')
+
+      else:
+        self.PublishMessage(f'Type {log_message_type} not supported')
+        resource = None
+
+      if resource:
+        if resource.id:
+          self.resources_dict[resource.id] = resource
+
+  def _ParseInsertLogMessage(self, resource: Resource, request: Dict[str, Any],
+                             response: Dict[str, Any]) -> None:
+    """Parses a GCP log message where the operation is insert or create.
+
+    Args:
+      resource: Resource object to update with parsed information
+      request: Request portion of the log message
+      response: Response portion of the log message
+    """
+    resource.creation_timestamp = response.get('insertTime')
+    resource.created_by = response.get('user', '')
+
+    if 'sourceDisk' in request:
+      if not resource.parent:
+        resource.parent = Resource()
+        resource.parent.resource_name = request.get('sourceDisk', '')
+
+    # When creating machine image from an instance
+    elif 'sourceInstance' in request:
+      if not resource.parent:
+        resource.parent = Resource()
+        resource.parent.resource_name = request.get('sourceInstance', '')
+
+    # When creating a disk image from another disk
+    elif 'sourceSnapshot' in request:
+      if not resource.parent:
+        resource.parent = Resource()
+        resource.parent.resource_name = request.get('sourceSnapshot', '')
```

Comment:
check the resource.parent once instead of three times
```suggestion
    if not resource.parent:
      if 'sourceDisk' in request:
        resource.parent = Resource()
        resource.parent.resource_name = request.get('sourceDisk', '')

      # When creating machine image from an instance
      elif 'sourceInstance' in request:
        resource.parent = Resource()
        resource.parent.resource_name = request.get('sourceInstance', '')

      # When creating a disk image from another disk
      elif 'sourceSnapshot' in request:
        resource.parent = Resource()
        resource.parent.resource_name = request.get('sourceSnapshot', '')
```

Reply:
what if there is no parent and none of these is in the request, should the function just continue or do you need to stop?

Reply:
I removed "if not resource.parent:" It was mainly there to initialise the resource object if it's null. I just initialise it anyway. And yes if no parent I still want to continue to parse disks.